cilium/cmd/kernel_check.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/spf13/cobra"
+)
+
+var kernelCheckCmd = &cobra.Command{
+	Use:   "kernel-check",
+	Short: "Checks whether machine is supporting required kernel capabilities",
+	Run: func(cmd *cobra.Command, args []string) {
+		kernelCheck()
+	},
+}
+
+func kernelCheck() {
+	probeManager := probes.NewProbeManager()
+	if err := probeManager.SystemConfigProbes(); err != nil {
+		log.WithError(err).Warning("BPF system config check: NOT OK")
```

Comment:
I think that this is probably a reasonable warning message for the CLI (/cc @cilium/bpf @mrostecki @qmonnet ):

```suggestion
		log.WithError(err).Warning("Unable to detect eBPF support. Is the kernel config available?")
```

Though I'm not deeply familiar with the details of in which cases this fails and whether we could for instance determine that *some* kernel config is available but the kernel in fact doesn't support eBPF (though most kernels do these days).

Reply:
Here, `err != nil` if one of the requirements to run Cilium isn't met. If [we fail to retrieve the config from bpftool](https://github.com/cilium/cilium/blob/6fc0b5b09d41dc4968290ced850da5c181847a40/pkg/datapath/linux/probes/probes.go#L140), we'll get `could not run bpftool` or `could not parse bpftool output`.

Reply:
I wonder if we could be a bit more selective here.

Consider if there was a `pkg/datapath/linux/probes/probes.go:GetRequiredConfig()` function like this:

```
func (p *ProbeManager) GetRequiredConfig() (supported map[KernelParam]bool) {
...
}
```

and inside this function it checked for all of the required config options that `SystemConfigProbes()` does today, and add them to a map with the key of the KernelParam and the value of whether it's enabled or not, and return that map. From `SystemConfigProbes()` we could call `GetRequiredConfig()` and iterate the map and if any of the required config options have the value `false`, create the error message and return it.

From `SystemConfigProbes()` there should be no real change in behaviour.

Then from this code, you could instead of calling `SystemConfigProbes()` call `GetRequiredConfig()` to get the map of support. Then to generate the commandline output, you iterate the map and print whether the feature is supported or not. 

If (and only if) all required parameters are set to false, then we could print the message "Is the kernel config available?" because in that case we don't appear to be getting any kind of signal from bpftool about supported configs.

(Separately, we could also add a `GetOptionalConfig()` which would get all of the options that the current `SystemConfigProbes()` only prints warnings for and doesn't return an error, the ones towards the end of the function; those could be also printed on this CLI).

Reply:
> [...] call GetRequiredConfig() and iterate the map and if any of the required config options have the value false, create the error message and return it. [...] 

@joestringer thanks for your input, I'm currently implementing this and would like to clarify about the supposed output of the CLI (see your comment above) when we are having a required config which is not employed/accesible on the respective node. 

```
# displays the config and whether it's enabled + appends error message
CONFIG_BPF_SYSCALL: Not supported -  Kernel parameter is required (mandatory) 
```

Reply:
Sorry, I came to this PR a couple of times but missed this question.

What I think probably makes sense is a table like this (can be achieved with `tabwriter`, grep for it in `cilium/` for examples):

```
FEATURE               SUPPORTED?    REQUIRED?
CONFIG_BPF_SYSCALL    true          true
...
```

Reply:
@joestringer output on our `Vagrant` box looks like this: 

```
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ sudo cilium kernel-check
FEATURE:   CONFIG_BPF_SYSCALL        SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_NET_SCH_INGRESS    SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_NET_CLS_BPF        SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_NET_CLS_ACT        SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_HAVE_EBPF_JIT      SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_BPF                SUPPORTED:true     REQUIRED: true
FEATURE:   CONFIG_BPF_EVENTS         SUPPORTED:true     REQUIRED: false
FEATURE:   CONFIG_CGROUP_BPF         SUPPORTED:false    REQUIRED: false
FEATURE:   CONFIG_LWTUNNEL_BPF       SUPPORTED:false    REQUIRED: false
```

Reply:
@joestringer not quite sure, whether it's possible with `tabwriter` to have only _one_ heading for each column as it loops to the whole map of course 🤔